File: CS/CustomizeExportAction.Module.Win/Controllers/CustomizeExportControllerWin.cs
using System;
using System.ComponentModel;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text;

using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Actions;
using DevExpress.Persistent.Base;

using DevExpress.ExpressApp.Win.SystemModule;
using DevExpress.ExpressApp.Win.Editors;
using DevExpress.ExpressApp.Win;
using System.Windows.Forms;

using System.IO;
using CustomizeExportAction.Module.Controllers;
using DevExpress.XtraGrid.Views.Grid;

namespace CustomizeExportAction.Module.Win.Controllers {
   public partial class CustomizeExportControllerWin : CustomizeExportController {
      private WinExportController winExportController;
      protected override void OnActivated() {
         base.OnActivated();
         winExportController = Frame.GetController<WinExportController>();
         winExportController.Exported += new EventHandler<DevExpress.ExpressApp.SystemModule.CustomExportEventArgs>(winExportController_Exported);
      }

      void winExportController_Exported(object sender, DevExpress.ExpressApp.SystemModule.CustomExportEventArgs e) {
         if (e.Stream is FileStream) {
            string fileName = ((FileStream)e.Stream).Name;
            if (File.Exists(fileName)) {
               e.Stream.Close();
               if (WinApplication.Messaging.GetUserChoice("Open the exported file?",
                   GetMessageBoxCaption(), MessageBoxButtons.YesNo) == DialogResult.Yes)
                  Process.Start(fileName);
            }
         }
      }
      protected override void CustomExport(object sender, DevExpress.ExpressApp.SystemModule.CustomExportEventArgs e) {
         base.CustomExport(sender, e);
         //Show a message before exporting a Grid List Editor
         GridListEditor gridListEditor =
             ((DevExpress.ExpressApp.ListView)View).Editor as GridListEditor;
         if (gridListEditor != null) {
            GridView gridView = gridListEditor.GridView;
            if (HasCollapsedGroups(gridView)) {
               string message =
                  "There are collapsed groups in the grid. Expand all groups in the exported file?";
               gridView.OptionsPrint.ExpandAllGroups =
                  WinApplication.Messaging.GetUserChoice(message, GetMessageBoxCaption(),
                  MessageBoxButtons.YesNo)
                  == DialogResult.Yes;
            }
         }
      }
      private bool HasCollapsedGroups(GridView gridView) {
         if (gridView.GroupCount > 0) {
            int rowHandle = -1;
            while (gridView.IsValidRowHandle(rowHandle)) {
               if (!gridView.GetRowExpanded(rowHandle))
                  return true;
               rowHandle--;
            }
         }
         return false;
      }
      private string GetMessageBoxCaption() {
         return String.Format(
             "{0} {1}", winExportController.ExportAction.Caption,
                 winExportController.ExportAction.SelectedItem);
      }
      protected override void OnDeactivated() {
         winExportController.Exported -= new EventHandler<DevExpress.ExpressApp.SystemModule.CustomExportEventArgs>(winExportController_Exported);
         base.OnDeactivated();
      }
   }
}

File: VB/CustomizeExportAction.Module.Win/Controllers/CustomizeExportControllerWin.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.ComponentModel
Imports System.Collections.Generic
Imports System.Diagnostics
Imports System.Text

Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Actions
Imports DevExpress.Persistent.Base

Imports DevExpress.ExpressApp.Win.SystemModule
Imports DevExpress.ExpressApp.Win.Editors
Imports DevExpress.ExpressApp.Win
Imports System.Windows.Forms

Imports System.IO
Imports CustomizeExportAction.Module.Controllers

Namespace CustomizeExportAction.Module.Win.Controllers
   Partial Public Class CustomizeExportControllerWin
	   Inherits CustomizeExportController
	  Private winExportController As WinExportController
	  Protected Overrides Sub OnActivated()
		 MyBase.OnActivated()
		 winExportController = Frame.GetController(Of WinExportController)()
		 AddHandler winExportController.Exported, AddressOf winExportController_Exported
	  End Sub

	  Private Sub winExportController_Exported(ByVal sender As Object, ByVal e As DevExpress.ExpressApp.SystemModule.CustomExportEventArgs)
		 If TypeOf e.Stream Is FileStream Then
			Dim fileName As String = (CType(e.Stream, FileStream)).Name
			If File.Exists(fileName) Then
			   e.Stream.Close()
			   If WinApplication.Messaging.GetUserChoice("Open the exported file?", GetMessageBoxCaption(), MessageBoxButtons.YesNo) = DialogResult.Yes Then
				  Process.Start(fileName)
			   End If
			End If
		 End If
	  End Sub
	  Protected Overrides Sub CustomExport(ByVal sender As Object, ByVal e As DevExpress.ExpressApp.SystemModule.CustomExportEventArgs)
		 MyBase.CustomExport(sender, e)
		 'Show a message before exporting a Grid List Editor
		 Dim gridListEditor As GridListEditor = TryCast((CType(View, DevExpress.ExpressApp.ListView)).Editor, GridListEditor)
		 If gridListEditor IsNot Nothing Then
			Dim gridView As XafGridView = gridListEditor.GridView
			If HasCollapsedGroups(gridView) Then
			   Dim message As String = "There are collapsed groups in the grid. Expand all groups in the exported file?"
			   gridView.OptionsPrint.ExpandAllGroups = WinApplication.Messaging.GetUserChoice(message, GetMessageBoxCaption(), MessageBoxButtons.YesNo) = DialogResult.Yes
			End If
		 End If
	  End Sub
	  Private Function HasCollapsedGroups(ByVal gridView As XafGridView) As Boolean
		 If gridView.GroupCount > 0 Then
			Dim rowHandle As Integer = -1
			Do While gridView.IsValidRowHandle(rowHandle)
			   If (Not gridView.GetRowExpanded(rowHandle)) Then
				  Return True
			   End If
			   rowHandle -= 1
			Loop
		 End If
		 Return False
	  End Function
	  Private Function GetMessageBoxCaption() As String
		 Return String.Format("{0} {1}", winExportController.ExportAction.Caption, winExportController.ExportAction.SelectedItem)
	  End Function
	  Protected Overrides Sub OnDeactivated()
		 RemoveHandler winExportController.Exported, AddressOf winExportController_Exported
		 MyBase.OnDeactivated()
	  End Sub
   End Class
End Namespace
